### ISSUE TITLE
Concurrency issue?

### ISSUE THREAD

**Issue by VoxBG** [simolus3/sqlite3.dart#31]
Probably after the last update of flutter (2.0.5), my unit tests started to fail, I notice something strange:
![image](https://user-images.githubusercontent.com/9137829/116614968-4c1ffd80-a943-11eb-88bc-253718fbc545.png)
As if the the select method is executed before the finish of the second execute method (teste it without transaction, the result is the same).
Unit test is executed in windows 10 environment with Sqlite.dll version 3.35.5.
Dart Sqlite3 version is 1.1.0.
Does anyone else have this issue or I am doing something wrong?

**Comment by simolus3** [comment]
I don't think it's a concurrency issue because sqlite3 apis are synchronous. Still, can you post the source code of the test so I can check if I can reproduce this?

**Comment by VoxBG** [comment]
You can add the following code to `sqlite3.dart\sqlite3\test\database_test.dart` and run the test:
``` DART
test('test', () {
    database.select('''
    CREATE TABLE Expense (
      id INTEGER NOT NULL PRIMARY KEY,
      createTimestamp INTEGER NOT NULL,
      updateTimestamp INTEGER NOT NULL,
      date INTEGER NOT NULL,
      description TEXT NOT NULL,
      value REAL NOT NULL
    );''');

    database.execute('''
    BEGIN TRANSACTION;
      ALTER TABLE Expense ADD COLUMN category_id INTEGER NOT NULL;
      ALTER TABLE Expense ADD COLUMN category_createTimestamp INTEGER NOT NULL;
      ALTER TABLE Expense ADD COLUMN category_updateTimestamp INTEGER NOT NULL;
      ALTER TABLE Expense ADD COLUMN category_name TEXT NOT NULL;
    COMMIT;''');

    var result = database.select('PRAGMA table_info(Expense);');
    print(result.rows.map((x) => x[1]));
  });
```
I am currently trying to recreated the scenario in C++ console application to see if the issue is from Sqlite3 itself.

**Comment by VoxBG** [comment]
The behavior is different in the c++ scenario, step for PRAGMA table_info statement returns error "cannot start a transaction within a transaction" which is strange because the step for the transaction statements returns SQL_DONE. I still can't figure out the reason for this.
I have attached the [code](https://github.com/simolus3/sqlite3.dart/files/6406980/sqlite_test.zip), I executed it in VS2019 with Desktop development with C++ workload, it includes also the executable binary.

**Comment by simolus3** [comment]
Interesting, when I change your C++ code to use `sqlite3_exec` to alter the table, the pragma statement does not find any rows. Interestingly, any other following select statement doesn't seem to find any rows either:

```c++
    //alter table
    string alterTableSql = "BEGIN TRANSACTION;\
        ALTER TABLE Expense ADD COLUMN category_id INTEGER NOT NULL;\
    ALTER TABLE Expense ADD COLUMN category_createTimestamp INTEGER NOT NULL;\
    ALTER TABLE Expense ADD COLUMN category_updateTimestamp INTEGER NOT NULL;\
    ALTER TABLE Expense ADD COLUMN category_name TEXT NOT NULL;\
    COMMIT; ";
    rc = sqlite3_exec(db, alterTableSql.c_str(), nullptr, nullptr, nullptr);
    fprintf(stdout, "done");
    if (rc != SQLITE_OK && rc != SQLITE_DONE) {
        fprintf(stderr, "SQL error: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        getchar();
        return(1);
    }

    //check table
    string tableInfoSql = "SELECT 1;";
    rc = sqlite3_prepare_v3(db, alterTableSql.c_str(),
        alterTableSql.length(), 0, &stmt, nullptr);
    if (rc != SQLITE_OK && rc != SQLITE_DONE) {
        fprintf(stderr, "SQL error: %s\n", sqlite3_errmsg(db));
        sqlite3_close(db);
        getchar();
        return(1);
    }

    while((rc = sqlite3_step(stmt)) == SQLITE_ROW) {
        fprintf(stdout, "row\n");
        const char* columnName = reinterpret_cast<const char*>(sqlite3_column_text(stmt, 1));
        printf(columnName);
        printf("\n");
    }
```

This library recently changed it's implementation of `database.execute`. In the past, we used `sqlite3_exec` as well, but we've now switched to using a prepared statement to support variables. Maybe that's a behavioral change you're seeing as well? Still, the behavior of the native code is surprising to me.

**Comment by VoxBG** [comment]
It is interesting indeed, why did it worked when the it does not in the c++ variant.

Also I manage to make my unit tests, which are related to the current issue, to pass when reverting back dart sqlite3 package to version 1.0.0.

And now with all this strange behavior I wonder if it is correct behavior the unit tests to pass and if there is a bug in the current version of dart sqlite3 package and if there are bugs in sqlite3 itself which cause inconsistent query results.

**Comment by VoxBG** [comment]
And yes it might be related that in version 1.0.0 execute method used sqlite3_exec and now it is using [prepared statement](https://github.com/simolus3/sqlite3.dart/commit/b82c9693ffa4b6fdf73acd4089cb0165aee431aa#diff-1dd356d3a6e2a095522e05600ac74f0b4de85c5b936a88baf8f8074aba923883).

**Comment by simolus3** [comment]
So I can actually reproduce this with a very minimal C example:

```c
#include <stdlib.h>
#include <stdio.h>
#include <sqlite3.h>

int main() {
	sqlite3* db;
	int rc;
	sqlite3_stmt* stmt;

	rc = sqlite3_open_v2(":memory:", &db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, 0);
	rc = sqlite3_exec(db, "CREATE TABLE foo (id INTEGER NOT NULL PRIMARY KEY)", 0, 0, 0);

	//rc = sqlite3_prepare_v3(db, "BEGIN TRANSACTION; ALTER TABLE foo ADD COLUMN bar INTEGER NOT NULL; COMMIT;", 75, 0, &stmt, 0);
	//rc = sqlite3_step(stmt);
	//sqlite3_finalize(stmt);
	sqlite3_exec(db, "BEGIN TRANSACTION; ALTER TABLE foo ADD COLUMN bar INTEGER NOT NULL; COMMIT;", 0, 0, 0); check;

	rc = sqlite3_prepare_v3(db, "PRAGMA table_info(foo);", 23, 0, &stmt, 0); check
	while ((rc = sqlite3_step(stmt)) == SQLITE_ROW) {
	    printf(sqlite3_column_text(stmt, 1));
	}
}
```

In this configuration, it prints `idbar`, if I use the prepared statement and comment out `sqlite3_exec`, it only prints `id`. I can't imagine that this is intended, I'll ask for advice on the sqlite3 forum.

**Comment by VoxBG** [comment]
Great, I will just use version 1.0.0 for now.

**Comment by simolus3** [comment]
Oh no wait, I think I figured it out. `sqlite3_prepare` can only prepare a SINGLE statement, it ignores the rest of the string after the first statement is parsed. So in your case it would just start the transaction. We can use the `const char **pzTail` argument to detect that.

I think what this library should do is:

- If no arguments are passed, use `sqlite3_exec`
- Otherwise, check that the provided SQL statement does not have trailing contents, and throw if it does.

**Comment by VoxBG** [comment]
This explains a lot and yes this is great idea, there should be a fallback to sqlite3_exec in order to be possible to execute multiple statements in one call.

**Comment by simolus3** [comment]
I've just published version 1.1.1 of the `sqlite3` package which has the new behavior.
